### PR TITLE
Creates dag op for a call of QASM custom gate

### LINKED ISSUE #1566: Qasms with composite gates don't load
Here is the qasm string

'OPENQASM 2.0;\ninclude "qelib1.inc";\ngate r q { h q; s q; h q; s q; }\ngate rinv q {sdg q; h q; sdg q; h q; }\ngate sinv q {s q; z q;}\nqreg q[3];\ncreg c[3];\nrinv q[0];\ncx q[1], q[2];\nr q[1];\nrinv q[2];\ncx q[0], q[2];\nrinv q[0];\nr q[2];\ncx q[0], q[1];\nrinv q[1];\ny q[0];\nbarrier q[0], q[1], q[2];\nr q[1];\ncx q[0], q[1];\nrinv q[2];\nr q[0];\ncx q[0], q[2];\nr q[2];\nrinv q[1];\ncx q[1], q[2];\nr q[0];\nx q[0];\nz q[1];\nmeasure q[0] -> c[0];\nmeasure q[1] -> c[1];\nmeasure q[2] -> c[2];'

If I run `from_qasm_str`

I get

`QiskitError: 'unknown operation for ast node name rinv'`

### Informations

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #1566

### CODE REVIEW

**Comment by ajavadia** [comment]
can you add a test and release note? thanks for taking on this old issue.

**Comment by 1ucian0** [comment]
I would like @kdk to have a look. It was too easy... I suspect that I cut corners :-/

**Comment by msegado** [comment]
Thanks for working on this. I'm not sure this fixes the issue, though - it seems to ignore the body from the gate definition and creates an empty instruction instead, which won't actually implement the operations defined in the QASM... right?

**Comment by msegado** [comment]
A simple(ish) workaround would be to unroll the custom gate rather than create a custom instruction. The algorithm would be look something like:
- When encountering a non-opaque custom gate call:
  - push its inputs onto the `arg_stack` and `bit_stack` (_Note: this is already implemented in `_process_custom_unitary()`_)
  - descend into the gate body and proceed with parsing and adding those operations directly (recursively if needed)
- When encountering an id (rather than a value) in a parameter or bit list:
  - look up and use the most recent value in the `arg_stack` and `bit_stack`
  - if no value is bound to the identifier in the stack, raise an exception

**Comment by 1ucian0** [comment]
> I'm not sure this fixes the issue, though 

You are right. Moving this to `[WIP]`

**Comment by 1ucian0** [comment]
This should be the test:
```python
from qiskit import QuantumCircuit, QuantumRegister
qc = QuantumCircuit.from_qasm_str('OPENQASM 2.0;\ninclude "qelib1.inc";\ngate rinv q {sdg q; h q; sdg q; h q; }\nqreg q[1];\nrinv q[0];')

rinv_q = QuantumRegister(1, name='q')
rinv_gate = QuantumCircuit(rinv_q, name='rinv')
rinv_gate.sdg(rinv_q)
rinv_gate.h(rinv_q)
rinv_gate.sdg(rinv_q)
rinv_gate.h(rinv_q)
rinv = rinv_gate.to_instruction() 

qr = QuantumRegister(1, name='qr') 
circuit = QuantumCircuit(qr, name='circuit') 
circuit.append(rinv_gate, [qr[0]])

print(circuit == qc)
```
The expected output is `True`.

**Comment by msegado** [comment]
That test seems reasonable, though at least for my use case even an unrolled version would be fine (e.g. if getting a composite instruction through the intermediate DAG representation proves difficult).

The following may be a helpful reference to base an approach on; this is how it used to be done in the older version of Qiskit: https://github.com/Qiskit/qiskit-terra/blob/829ca256942495faedb705bda718d641da8e4c63/qiskit/unroll/_unroller.py#L83-L108

**Comment by 1ucian0** [comment]
I think I now have a proper fix. Have a look!

**Comment by msegado** [comment]
Thanks again for all your work on this! I hate to bring bad news, but I think there are still a few issues with this version:

1. It assumes all child gates have the same input bits as the parent, and in the same order; i.e., the following gate definitions in the standard library will parse to incorrect instructions:
```c
gate swap a,b {
  cx a,b;
  cx b,a;  // issue: different bit order
  cx a,b;
}
gate cswap a,b,c
{
  cx c,b;  // issue: different bit count
  ccx a,b,c;
  cx c,b;
}
```
2. It assumes (I think?) that the child instructions aren't parametrized; also from the standard library:
```c
gate u2(phi,lambda) q {
  U(pi/2,phi,lambda) q;  // issue: parameters ignored/not mapped
}
```
I imagine one could write tests for those, though I'm not 100% sure how the `Parameter`s work in circuits or how/if parameter expressions are compared for equality...

**Comment by 1ucian0** [comment]
Thanks a lot @msegado for you comments and cases! Have another look and let me know if you find more!

**Comment by msegado** [comment]
@1ucian0 Very welcome! And thanks again for all your efforts in implementing this.

I was playing around and found a new test case — seems something breaks when bound parameters are used in math expressions:

```c
OPENQASM 2.0;
include "qelib1.inc";
gate my_u2(phi,lambda) q {u2(phi+pi,lambda/2) q;} // parameters used in expressions
qreg qr[1];
my_u2(pi, pi) qr[0];
```
Trying to load this with `QuantumCircuit.from_qasm_str` throws a `NodeException` with the following stack trace:
```
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
  File "/app/src/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 1240, in from_qasm_str
    return _circuit_from_qasm(qasm)
  File "/app/src/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 1321, in _circuit_from_qasm
    dag = ast_to_dag(ast)
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 93, in ast_to_dag
    AstInterpreter(dag)._process_node(ast)
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 276, in _process_node
    self._process_children(node)
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 271, in _process_children
    self._process_node(kid)
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 311, in _process_node
    self._process_custom_unitary(node)
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 203, in _process_custom_unitary
    [self.bit_stack[-1][s] for s in gbits])
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 402, in _create_dag_op
    op = self._create_op(name, params)
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 419, in _create_op
    params=params)
  File "/app/src/qiskit-terra/qiskit/converters/ast_to_dag.py", line 385, in _gate_definition_to_definition
    eparams.append(param.sym())
  File "/app/src/qiskit-terra/qiskit/qasm/node/binaryop.py", line 60, in sym
    lhs = self.children[1].sym(nested_scope)
  File "/app/src/qiskit-terra/qiskit/qasm/node/id.py", line 68, in sym
    self.name, self.line, self.file))
qiskit.qasm.node.nodeexception.NodeException: 'Expected local parameter name:  name=phi, line=3, file='
```
Haven't looked into it in enough detail to identify the root issue though...

File: qiskit/converters/ast_to_dag.py [hunk 1]
```diff
@@ -361,6 +362,31 @@ def _process_node(self, node):
                               "file=%s" % node.file)
         return None
 
+    def _gate_definition_to_definition(self, node, params):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/converters/ast_to_dag.py]: Done in 11571618d

**Comment by 1ucian0** [comment]
Thanks @msegado !

I addressed it in d93128cba. Any other case?

**Comment by msegado** [comment]
@1ucian0 Nothing else on my end so far!

**Comment by 1ucian0** [comment]
> @1ucian0 Nothing else on my end so far!

Great! Thanks a lot again! @kdk @ajavadia , shall we merge?

**Comment by msegado** [comment]
@1ucian0 Oops, spoke too soon; expressions with functions still fail with a NodeException for me. Here's a modified test case that catches it:

```c
OPENQASM 2.0;
include "qelib1.inc";
gate my_u2(phi,lambda) q {u3(asin(cos(phi)/2), phi+pi, lambda/2) q;} // parameters used in expressions
qreg qr[1];
my_u2(pi, pi) qr[0];
```

**Comment by itoko** [comment]
@1ucian0 Thank you for tackling this, I've really waited for this fix!

I just want to confirm the scope of this fix. 

1) It works for nesting custom gates (without cycle), right? e.g.
```
gate child a,b {
  ...
}
gate parent a,b {
  ...
  child a, b;
  ...
}
```

2) Definition order matters? For example, does this work?
```
gate parent a,b {
  ...
  child a, b;
  ...
}
gate child a,b {
  ...
}
```

3) Can we write the loaded QASM back as the same QASM (with the custom definitions)?

Excuse me for not checking these by running the code by myself.
I don't mean all of these should be covered in this PR (I think it's still very useful even if none of them are covered). I just want to let the user know the details of what they can/cannot with the new feature.

**Comment by ajavadia** [comment]
Can you do a quick check that the following qasm files load?

https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/generic/bigadder.qasm

https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/generic/qpt.qasm

https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/generic/pea_3_pi_8.qasm

https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/generic/adder.qasm

https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/generic/ipea_3_pi_8.qasm

https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/generic/teleportv2.qasm

https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/generic/qec.qasm

And this one should error:
https://github.com/Qiskit/openqasm/blob/5fb61893661261b7211942dc04ee85f8fd845f2a/examples/invalid/gate_no_found.qasm

**Comment by 1ucian0** [comment]
@msegado said:
> @1ucian0 Oops, spoke too soon; expressions with functions still fail with a NodeException for me.

I made my fix much more generic. It should be solved now. Have a look!

**Comment by 1ucian0** [comment]
@itoko said:
> 1. It works for nesting custom gates (without cycle), right? e.g.

yes. Testing that in 9e62d2c0c

> 2. Definition order matters?

Yes, order matters. It seems to be in the spirit of the specification... Your example does not work.

> 3. Definition order matters? For example, does this work?

You mean `circuit.qasm()`? Let's leave the exporter out of this PR for now...

**Comment by 1ucian0** [comment]
@ajavadia said:
> Can you do a quick check that the following qasm files load?

Yes. They all load (and the error with the last one is coming from the parser). I also transpiled these circuits once they were loaded (with `transpile(QuantumCircuit.from_qasm_file('/file), basis_gates=['u1', 'u2', 'u3', 'cx'])` and no errors.

File: releasenotes/notes/from_qasm_str_custom_gates-e3013f958def70e8.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+---
+fixes:
+  - |
+    A bug in ``from_qasm_str`` loading custom gates was fixed. Now, loading this QASM:
```
Comment (mtreinish) [inline releasenotes/notes/from_qasm_str_custom_gates-e3013f958def70e8.yaml]:
```suggestion
    A bug in ``from_qasm_str`` loading custom gates was fixed. Now, loading this QASM::
```

File: releasenotes/notes/from_qasm_str_custom_gates-e3013f958def70e8.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+        OPENQASM 2.0;
+        include "qelib1.inc";
+        gate rinv q {sdg q; h q; sdg q; h q; }
+        qreg q[1];
+        rinv q[0];"""
+     
+    is equivalent to the following circuit:
```
Comment (mtreinish) [inline releasenotes/notes/from_qasm_str_custom_gates-e3013f958def70e8.yaml]:
```suggestion
    is equivalent to the following circuit::
```

**Pull request merged**